urllib3/util/connection.py

```diff
@@ -57,6 +61,12 @@ def create_connection(address, timeout=socket._GLOBAL_DEFAULT_TIMEOUT,
     # The original create_connection function always returns all records.
     family = allowed_gai_family()
 
+    # If IPv6 and selectors are available, use the Happy Eyes algorithm.
+    # (RFC 6555 https://tools.ietf.org/html/rfc6555)
+    if HAS_IPV6 and HAS_SELECT and ENABLE_HAPPY_EYEBALLS:
+        return happy_eyeballs_algorithm((host, port), timeout,
+                                        source_address, socket_options)
```

Comment:
I'm not sure about bringing it into the same file, I'm just suggesting that we should take a very close look to see if there is common code that can be factored out or used in both places.